### ISSUE TITLE
No module named 'rlpyt.samplers.cpu'

### ISSUE THREAD

**Issue by tzahishimkin** [astooke/rlpyt#15]
Running 'mujoco_ff_a2c_cpu.py' and getting this error. Can't find 'cpu' directory in 'samples', only inside the subdirectory 'parallel'

**Comment by tzahishimkin** [comment]
I might be missing on something, but I am unable to locate the following libraries:

from rlpyt.samplers.gpu.parallel_sampler import GpuParallelSampler
from rlpyt.samplers.gpu.collectors import WaitResetCollector
from rlpyt.runners.minibatch_rl_eval import MinibatchRlEval

**Comment by mlpanda** [comment]
Just adding to this comment; I tried running the r2d1 experiment for Atari, and constantly ran into functions which did not exist. Initially I debugged and found that it was mainly due to changes in function names, however at some stage I ran into larger changes, hence this comment. @astooke I would very much appreciate if you could look through the code and update the experiments to new changes elsewhere in the repo - and perhaps add an r2d1 example file :) Thanks in advance.

**Comment by astooke** [comment]
Hi!  Yes indeed, the experiments folder has a bunch of leftover scripts written over the course of development, as things were changing.  Definitely that should be cleaned up, to leave the only the current running examples.  Will do soon!  Sorry about the confusion.

**Comment by mlpanda** [comment]
Thanks @astooke, much appreciated! :-)

**Comment by tzahishimkin** [comment]
Thanks allot!

So which example code do you use to run the DDPG model?

**Comment by astooke** [comment]
Hi! Sorry lost this for a while...but a current working example should be:

https://github.com/astooke/rlpyt/blob/master/rlpyt/experiments/scripts/mujoco/qpg/launch/got/launch_mujoco_ddpg_serial.py

which calls

https://github.com/astooke/rlpyt/blob/master/rlpyt/experiments/scripts/mujoco/qpg/train/mujoco_ddpg_serial.py

**Comment by DanielTakeshi** [comment]
(I think the launch script you're referring to shouldn't be in the `got` subdirectory -- incidentally what does that name represent?)

@astooke  I have several quick related questions.

I am trying to understand the experiments directory. These are serial as judged by the file names, and I see the serial sampler here:

https://github.com/astooke/rlpyt/blob/75e96cda433626868fd2a30058be67b99bbad810/rlpyt/experiments/scripts/mujoco/qpg/train/mujoco_ddpg_serial.py#L24-L30

However, the launch file has this:

https://github.com/astooke/rlpyt/blob/75e96cda433626868fd2a30058be67b99bbad810/rlpyt/experiments/scripts/mujoco/qpg/launch/launch_mujoco_ddpg_serial.py#L7-L13

I'm a bit confused because serial implies we have one python process running everything. However, the launch file suggests we can have lots of CPUs and lots of GPUs used. However this would suggest to me using multiple processes. I am wondering, what is the proper way to interpret `n_cpu_core` and `n_gpu_core` in this context?

**Comment by DanielTakeshi** [comment]
Just a quick note that none of the four current policy gradient scripts for MuJoCo appear to be up to date with the latest changes on master today:

![Screenshot from 2019-11-21 09-34-50](https://user-images.githubusercontent.com/5532096/69361930-5617da00-0c42-11ea-914c-af51a28e6804.png)

**Comment by DanielTakeshi** [comment]
I finally  managed to figure out how to parallelize the DDPG for MuJoCo,  because only the serial ones were available. Here is a working launch file for the CPU:


and the corresponding training script:


Note that these require changes into the config to create some new dictionaries:

https://github.com/astooke/rlpyt/blob/master/rlpyt/experiments/configs/mujoco/qpg/mujoco_ddpg.py
# Daniel: just add a bunch more keys.
config = copy.deepcopy(config)
config['sampler'].pop('batch_B')
config['sampler'].pop('batch_T')
#config['sampler']['batch_T'] = 1   # Actually let's just do this in the launch file.
#config['sampler']['batch_B'] = 10  # Same comment as above
configs['ddpg_from_td3_1M_parallel_cpu'] = config
config = copy.deepcopy(config)
configs['ddpg_from_td3_1M_parallel_gpu'] = config

A similar pattern can be done for the GPU.

**Comment by DanielTakeshi** [comment]
@astooke  I see how the experiments directory is generally structured. Do you have some suggestions or hacks you developed along the way to improve workflow? Mainly I am wondering because it is a lot of typing and hitting TAB to get to the files deep into the repository (the launch and training files).

**Comment by astooke** [comment]
Yes the general workflow included three files: 1) a "run" script which builds multiple hyperparameter settings and launches a separate experiment for each one (the one script to manually call), 2) a "train" script which builds one individual experiment and trains the agent, 3) optionally a "config" file which can store all the default settings for a given group of experiments, so that only the hyperparameters being swept need to be changed in the "run" script, since the "train" script has access to the config file.  

I'll keep all of those under one sub-directory in the experiments folder.  Then yeah after building up many of those when working along one line of research, eventually I'll start a new folder to work inside.  

Definitely not the only way to do it, but after doing another project or two using rlpyt, I've stuck with that.  :)

**Comment by astooke** [comment]
OK there are still a bunch of extra launch scripts floating around, but they basically just correspond to different computer configurations, and I think they all run now (or have their own issue), so will leave it as is for the time being.  :)   Please reopen or make an issue if you see a specific one you want that is broken.

**Comment by DanielTakeshi** [comment]
Minor note, the Categorical DQN scripts:

https://github.com/astooke/rlpyt/blob/master/rlpyt/experiments/scripts/atari/dqn/train/atari_catdqn_gpu.py
https://github.com/astooke/rlpyt/blob/master/rlpyt/experiments/scripts/atari/dqn/launch/launch_atari_catdqn_gpu_basic.py

still use the outdated paths. These can be replaced with this git diff:

```
-from rlpyt.samplers.gpu.sampler import GpuSampler
-from rlpyt.samplers.gpu.collectors import GpuWaitResetCollector
+from rlpyt.samplers.parallel.gpu.sampler import GpuSampler
+from rlpyt.samplers.parallel.gpu.collectors import GpuWaitResetCollector
```

**Comment by astooke** [comment]
@DanielTakeshi Thanks, just fixed that last one!
